R/easystats_downloads.R

```diff
@@ -0,0 +1,177 @@
+  downloads_all <- do.call("rbind", downloads_list)
+
+  # Helper function to calculate months between two dates
+  .monnb <- function(d) {
+    lt <- as.POSIXlt(as.Date(d, origin = "1900-01-01"))
+    lt$year * 12 + lt$mon
+  }
+
+  .mondf <- function(d1, d2) {
+    .monnb(d2) - .monnb(d1)
+  }
+
+  # Calculate total and monthly downloads per package
+  stats <- lapply(split(downloads_all, downloads_all$package), function(d) {
+    tmp <- d[d$count > 0, ]
+    if (nrow(tmp) == 0) {
+      return(data.frame(
+        package = d$package[1],
+        total = 0,
+        monthly = 0,
+        stringsAsFactors = FALSE
+      ))
+    }
+
+    months_on_cran <- .mondf(tmp$date[1], tmp$date[nrow(tmp)])
+    if (!length(months_on_cran) || months_on_cran < 1) {
+      months_on_cran <- 1
+    }
+
+    data.frame(
+      package = d$package[1],
+      total = sum(tmp$count),
+      monthly = round(sum(tmp$count) / months_on_cran),
+      stringsAsFactors = FALSE
+    )
+  })
+
+  out <- do.call(rbind, stats)
+  rownames(out) <- NULL
+
+  # Sort data
+  if (!is.null(sort_by)) {
+    sort_by <- insight::validate_argument(
+      sort_by,
+      c("package", "total", "monthly")
+    )
+    out <- out[order(out[[sort_by]], decreasing = TRUE), ]
+  }
+
+  # Shorten long strings
+  if (!is.null(length)) {
+    out$package <- insight::format_string(out$package, length)
+  }
+
+  # Add total row
+  out <- rbind(
+    out,
+    data.frame(
+      package = "Total",
+      total = sum(out$total),
+      monthly = sum(out$monthly),
+      stringsAsFactors = FALSE
+    )
+  )
+
+  # Capitalize column names
+  colnames(out) <- tools::toTitleCase(colnames(out))
+  class(out) <- c("easystats_downloads", "data.frame")
+
+  out
+}
+
+
+# methods ------------------------------------------------------------------
+
+#' @export
+print.easystats_downloads <- function(x, ...) {
+  # Format numeric columns with thousand separators
+  x_formatted <- x
+  x_formatted$Total <- format(x$Total, big.mark = ",", scientific = FALSE)
+  x_formatted$Monthly <- format(x$Monthly, big.mark = ",", scientific = FALSE)
+
+  cat(insight::export_table(x_formatted, ...))
+}
+
+#' @export
+print_html.easystats_downloads <- function(x, ...) {
+  # Format numeric columns with thousand separators
+  x_formatted <- x
+  x_formatted$Total <- format(x$Total, big.mark = ",", scientific = FALSE)
+  x_formatted$Monthly <- format(x$Monthly, big.mark = ",", scientific = FALSE)
+
+  insight::export_table(x_formatted, format = "html", ...)
+}
+
+#' @export
+print_md.easystats_downloads <- function(x, ...) {
+  # Format numeric columns with thousand separators
+  x_formatted <- x
+  x_formatted$Total <- format(x$Total, big.mark = ",", scientific = FALSE)
+  x_formatted$Monthly <- format(x$Monthly, big.mark = ",", scientific = FALSE)
+
+  insight::export_table(x_formatted, format = "markdown", ...)
+}
+
+#' @export
+display.easystats_downloads <- function(object, format = "markdown", ...) {
```

Comment:
We could add a `"tt"` option and include it like in other packages:

https://github.com/easystats/modelbased/blob/main/R/display.R

and see `.check_format_backend()`: https://github.com/easystats/modelbased/blob/main/R/print_html.R

Reply:
I think we can wait on merging easystats/insight#1137, and then switch to `insight::export_table()` here with `big_mark = ","` by default? Will reduce code redundancy and then format with all options can be directly specified there.